### ISSUE TITLE
[clang] False Positive Wdeclaration-after-statement

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#53438]
The following code triggers a False Positive:

```c
int square(int num) {
    int x = 123;
    x = 321;
    int y = 456;
    return x * y;
}
```

``` 
$ clang -std=c99 -Wdeclaration-after-statement -c main.c
warning: mixing declarations and code is incompatible with standards before C99 [-Wdeclaration-after-statement]
    int y = 456;
        ^
```

Example in [Compiler Explorer](https://godbolt.org/z/94GazPbns).

The warning applies to standards **before** C99, but I am using C99 in the code.

**Comment by zero9178** [comment]
It used to be the case that this warning was disabled when not compiling in versions older than C99 but users requested it to still warn if they explicitly ask for it as you are doing here. See https://bugs.llvm.org/show_bug.cgi?id=51931. It was therefore later enabled for all C versions, but disabled by default: https://reviews.llvm.org/D114787 and https://reviews.llvm.org/D117232. This also matches GCC behaviour. 

Is there any reason as to why you'd want it not to warn despite explicitly asking for the warning?

**Comment by carlosgalvezp** [comment]
Thanks for the input! I understand the rationale better now, I got confused because the warning message only talks about standards before C99. It would be good maybe to mention that this warning is intended for backwards compatibility. Makes sense to keep it consistent with GCC.

> Is there any reason as to why you'd want it not to warn despite explicitly asking for the warning?
I got this warning via `-Weverything`. Now that I know the purpose I can safely just disable this particular warning.
